module/epochs/epoch_lookup_test.go

```diff
@@ -1,65 +1,255 @@
+
+	// wait for the protocol event to be processed (async)
+	assert.Eventually(suite.T(), func() bool {
+		_, err := suite.lookup.EpochForViewWithFallback(suite.currEpoch.finalView + 1)
+		return err == nil
+	}, 5*time.Second, 50*time.Millisecond)
+
+	// validate queries are answered correctly
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, suite.currEpoch)
+
+	// should handle multiple deliveries of the protocol event
+	suite.lookup.EpochEmergencyFallbackTriggered()
+	suite.lookup.EpochEmergencyFallbackTriggered()
+	suite.lookup.EpochEmergencyFallbackTriggered()
+
+	// validate queries are answered correctly
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, suite.currEpoch)
+}
+
+func (suite *EpochLookupSuite) TestProtocolEvents_CommittedEpoch() {
+	// initially, only current epoch is committed
+	suite.CommitEpochs(suite.currEpoch)
+	suite.CreateAndStartEpochLookup()
+
+	// commit the next epoch, and emit a protocol event
+	firstBlockOfCommittedPhase := unittest.BlockHeaderFixture()
+	suite.state.On("AtBlockID", firstBlockOfCommittedPhase.ID()).Return(suite.snapshot)
+	suite.CommitEpochs(suite.nextEpoch)
+	suite.lookup.EpochCommittedPhaseStarted(suite.currentEpochCounter, firstBlockOfCommittedPhase)
+
+	// wait for the protocol event to be processed (async)
+	assert.Eventually(suite.T(), func() bool {
+		_, err := suite.lookup.EpochForViewWithFallback(suite.currEpoch.finalView + 1)
+		return err == nil
+	}, 5*time.Second, 50*time.Millisecond)
+
+	// validate queries are answered correctly
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, suite.currEpoch, suite.nextEpoch)
+
+	// should handle multiple deliveries of the protocol event
+	suite.lookup.EpochCommittedPhaseStarted(suite.currentEpochCounter, firstBlockOfCommittedPhase)
+	suite.lookup.EpochCommittedPhaseStarted(suite.currentEpochCounter, firstBlockOfCommittedPhase)
+	suite.lookup.EpochCommittedPhaseStarted(suite.currentEpochCounter, firstBlockOfCommittedPhase)
+
+	// validate queries are answered correctly
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, suite.currEpoch, suite.nextEpoch)
+}
+
+// testEpochForViewWithFallback accepts a constructed EpochLookup and state, and
+// validates correctness by issuing various queries, using the input state and
+// epochs as source of truth.
+func testEpochForViewWithFallback(t *testing.T, lookup *EpochLookup, state protocol.State, epochs ...epochRange) {
+	epochFallbackTriggered, err := state.Params().EpochFallbackTriggered()
+	require.NoError(t, err)
+
+	t.Run("should have set epoch fallback triggered correctly", func(t *testing.T) {
+		assert.Equal(t, epochFallbackTriggered, lookup.epochFallbackIsTriggered.Load())
+	})
+
+	t.Run("should be able to query within any committed epoch", func(t *testing.T) {
+		for _, epoch := range epochs {
+			counter, err := lookup.EpochForViewWithFallback(unittest.Uint64InRange(epoch.firstView, epoch.finalView))
```

Comment:
this will sample a random view from the entire epoch length. I think the chances to sample the first or last view are comparatively small. So a bug in the implementation that only affects the last or first view would show up rarely.

Suggestion: query the Epoch for 3 view numbers: the first view, the last view and a random view in between

Reply:
Good point, updated in https://github.com/onflow/flow-go/pull/2927/commits/e191036af575ed1c695a97d7819206552d9ac9ed